config/initializers/action_dispatch_url_helper.rb

```diff
@@ -0,0 +1,24 @@
+module ActionDispatch
+  module Routing
+    class RouteSet
+      class NamedRouteCollection
+        class UrlHelper
+          def initialize(route, options, route_name)
+            @options      = options
+            # Hack to prevent :locale being passed as the first missing positional argument,
+            # since we cannot explicitly set default value of :locale param in routes
+            # due to a Rails bug since 4.0 on default value of optional scope.
+            # As this might involve some bigger architecture issue of ActionDispatch(?),
+            # an official fix is not likely to come anytime soon.
+            # See: https://github.com/rails/rails/issues/12178
+            # See: https://github.com/rails/rails/pull/39582
```

Comment:
I'm going to read up on the linked references, thanks for including those.

Reply:
FYI during the research I also found some people blaming Rspec (https://github.com/rspec/rspec-rails/issues/255, https://github.com/rspec/rspec-rails/issues/562) for similar symptoms but I'm not sure if they are because my problem reproduces on the development env as well.

Reply:
I'm still investigating, but as part of that investigation, I spun up a brand-new Rails app to see how it behaved. You [can follow along here](https://github.com/jaw6/RailsSeven/blob/default_url/app/views/whatever/show.html.erb) if you like, but so far, it seems like the `default_url_options` method works as expected without any additional supports. Porting that over to `forem` does _not_ work, so there's something about our config that's doing this. (Still not sure if this will turn up anything useful, but thought I'd share what I have so far, in case it's enlightening.)

Reply:
Okay, still feeling like I have an incomplete understanding of this, but want to summarize what I've found so far:

* With `defaults: { locale: nil }` (as is being set in [`config/routes.rb`](https://github.com/forem/forem/blob/main/config/routes.rb#L25)), links using Rails helpers won't have "sticky locale" (regardless of whatever happens in `ApplicationController#default_url_options`) (in other words, locale switch won't be preserved in later clicks, which is not great).
* Without `defaults: { locale: nil }`, named url/path helpers with positional arguments (for example: `tag_path(tag)`) will throw exceptions, as they attempt to match the supplied object to path segments _starting with_ `:locale` and not having enough arguments to match `:id` (or `:tag` or what-have-you).
* This initializer overrides the named route generation at `draw` time, so any routes with a `:locale` segment (which, in effect, is _most_ of the front-end routes for forem) effectively _ignore_ the `locale` segment when "counting" segments to apply towards positional arguments, in effect allowing us to have both "sticky routes" and "named routes" _without_ having to update path/url callsites to explicitly manage locale, without effecting other route scopes / segment keys.
* The main downside seems to be that if any future Rails upgrade introduces changes in this part of the routing system, we have a somewhat brittle monkey-patch to sort out, potentially slowing down or even blocking future Rails upgrades.

Reply:
Thank you for your elaboration that I might have had to do. I think we are basically on the same page.

To justify why I took this approach, I think this part of code is well integrated to the whole logic (it's taken from [this file](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/routing/route_set.rb)) and no drastic rewiring is expected to happen. But indeed there will be non-zero possibility of update. I believe its performance hit is minimal as it only works at the path helpers generation timing on the start up.

Searching Rails' issue list led me to this thread https://github.com/joemasilotti/railsdevs.com/pull/256, it seems that they fell in the same pitfall as ours, and while I didn't read through the discussion, their final solution is to make a custom url helper and include it anywhere in place of the original one. This could be a more modest way, do you think it is more feasible?